### DISCUSSION TITLE
How to authenticate to a private GitHub repo.

### DISCUSSION THREAD

**Discussion by mc1903** [semaphoreui/semaphore#1640] [General]
Running Semaphore v2.9.37 and I cannot get it to authenticate to a private GitHub repo.

The docs are light on detail when it comes to this. I have been trying to get this working with zero success. If anyone has a walk through showing the configuration steps on both the Semaphore and GitHub sides, I would be grateful if you could share.

- What keypair algorithm(s) are supported for the public/private key pairs? RSA (size) / ed25519
- Is a pass-phrase on the private key supported or not?
- Where in GitHub is the SSH key added? The 'global' account settings for SSH keys (assume it would be an 'authentication key') or is it in the actual repo's Deploy Keys?
- should I use the GitHub repo's ssh or https git URL?

Any & all help appreciated.

Cheers,
M

**Comment by mc1903** [comment]
So, after a few more hours of banging my head against the keyboard, I finally have it working; so I can now answer my own questions. 

> What keypair algorithm(s) are supported for the public/private key pairs? RSA (size) / ed25519

Both RSA (4096 bits) and ed25519 algorithms work.

> Is a pass-phrase on the private key supported or not?

Yes & Yes. v2.9.37 does NOT support a passphrase on the private key, but v2.9.39-beta DOES support it.

> Where in GitHub is the SSH key added? The 'global' account settings for SSH keys (assume it would be an 'authentication key') or is it in the actual repo's Deploy Keys?

I added my public keys into the Repository's Deploy Keys as Read Only (or rather I did not check the 'Allow write access').

> should I use the GitHub repo's ssh or https git URL?

ssh works - but there is a very annoying catch. 

Paste in the git url `git@github.com:mc1903/Ansible.git` but **DO NOT** touch the little blue 'ssh' icon, as it will prepend `ssh://` onto the git url and this makes it fail.


![Ansible Seamphore - Edit Repo - Not Working](https://github.com/ansible-semaphore/semaphore/assets/13692287/695d690b-81ea-4d95-8696-f1bae3e4d9c2)
**This will fail**


![Ansible Seamphore - Edit Repo - Working](https://github.com/ansible-semaphore/semaphore/assets/13692287/7363a5a9-c3c6-4716-9c67-510fd30e559f)
**This works**

I had a combination of issues - I was using a passphrase on the private key with v2.9.37 and I had inadvertently touched the blue 'ssh' icon.

A BIG thanks to my buddies tcpdump and Wireshark for your help today.

I hope this helps someone else avoid a bad day!

**Reply by nicedevil007** [reply]
how did you get it work?

I created multiple times a SSH Key of type ed25519 WITH a passphrase
But semaphore is always telling me on git clone part this:
```
Failed updating repository: parsing private key: ssh: key is not password protected
```

and yes I'm on latest beta build so v2.9.39

![image](https://github.com/ansible-semaphore/semaphore/assets/17103076/bf78b594-1c36-411b-8509-6ca5fe143a38)

**Reply by nicedevil007** [reply]
I further investigated this and found out that creating the key in AlmaLinux 9.3 with OpenSSH_8.7p1, OpenSSL 3.0.7 1 Nov 2022 was doing something else than creating a key that works with ansible and gitea.

The command I use to create a ed25519 key is the following:
```powershell
ssh-keygen -o -a 255 -t ed25519 -f ~/.ssh/testKey -C "testingKey"
```

I created the keys now with my windows powershell 7.4 and then it is working like a charm. 2 frustrating hours of finding this out.

**Reply by mc1903** [reply]
Glad you got it working in the end.
